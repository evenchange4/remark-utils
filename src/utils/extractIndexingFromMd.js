// @flow
import remark from 'remark';
import remarkGridTables from 'remark-grid-tables';
import remarkHtml from 'remark-html';
import remarkPrismjs from 'gatsby-remark-prismjs';
import remarkAutolink from 'gatsby-remark-autolink-headers';
import remarkStrip from 'remark-strip-html';
import { type IndexingItem, type Ast } from './type.flow';
import defaultOptions from './defaultOptions';
import remarkIndexing from './remark-indexing';

const extractIndexingFromMd = (
  markdownContent: string,
): Array<IndexingItem> => {
  let indexingItems: Array<IndexingItem> = [];

  remark()
    .use(remarkGridTables)
    .use(() => (markdownAST: Ast) =>
      remarkAutolink({ markdownAST }, { icon: '', classNam: 'anchor' }),
    )
    .use(() => (markdownAST: Ast) =>
      remarkPrismjs({ markdownAST }, defaultOptions.prismjs),
    )
    .use(remarkStrip)
    .use(() => (markdownAST: Ast) => {
      indexingItems = remarkIndexing(markdownAST);
      return markdownAST;
    })
    .use(remarkHtml)
    .processSync(markdownContent);

  return indexingItems;
};

export default extractIndexingFromMd;
