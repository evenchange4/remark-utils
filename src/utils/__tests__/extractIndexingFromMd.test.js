// @flow
import extractIndexingFromMd from '../extractIndexingFromMd';

it('extractIndexingFromMd', () => {
  expect(
    extractIndexingFromMd(`
# header 1 標頭
1
1-1
1-2
## header 2 標頭
2

### header 3 標頭
3

#### header 4 標頭
4

##### header 5 標頭
5
5-1
###### header 6 標頭
6
6-1
`),
  ).toMatchSnapshot();
});
