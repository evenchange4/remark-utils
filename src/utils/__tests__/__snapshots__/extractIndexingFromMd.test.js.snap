// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractIndexingFromMd 1`] = `
Array [
  Object {
    "body": " header 1 標頭 1
1-1
1-2",
    "title": "header 1 標頭",
    "url": "#header-1-標頭",
  },
  Object {
    "body": " header 2 標頭 2",
    "title": "header 2 標頭",
    "url": "#header-2-標頭",
  },
  Object {
    "body": " header 3 標頭 3",
    "title": "header 3 標頭",
    "url": "#header-3-標頭",
  },
  Object {
    "body": " header 4 標頭 4",
    "title": "header 4 標頭",
    "url": "#header-4-標頭",
  },
  Object {
    "body": " header 5 標頭 5
5-1",
    "title": "header 5 標頭",
    "url": "#header-5-標頭",
  },
  Object {
    "body": " header 6 標頭 6
6-1",
    "title": "header 6 標頭",
    "url": "#header-6-標頭",
  },
]
`;
